python/tests/test_utils.py

```diff
@@ -0,0 +1,71 @@
+from copy import copy
+
+from numpy.testing import assert_allclose
+
+import rascaline
+from rascaline.systems import IntoSystem
+
+
+def finite_differences_positions(
+    calculator: rascaline.CalculatorBase,
+    system: IntoSystem,
+    displacement: float = 1e-6,
+    rtol: float = 1e-5,
+    atol: float = 1e-7,
+) -> None:
+    """
+    Check that analytical gradients with respect to positions agree with a finite
+    difference calculation of the gradients.
+
+    The implementation is simular to ``rascaline/src/calculators/tests_utils.rs``.
+
+    :param calculator: calculator used to compute the representation
+    :param system: Atoms object
+    :param displacement: distance each atom will be displaced in each direction when
+        computing finite differences
+    :param max_relative: Maximal relative error. ``10 * displacement`` is a good
+        starting point
+    :param atol: Threshold below which all values are considered zero. This should be
+        very small (1e-16) to prevent false positives (if all values & gradients are
+        below that threshold, tests will pass even with wrong gradients)
+    :raises AssertionError: if the two gradients are not equal up to specified precision
+    """
+    reference = calculator.compute(system, gradients=["positions"])
+
+    for atom_i in range(len(system)):
+        for spatial in range(3):
+            system_pos = copy(system)
+            system_pos._positions[atom_i][spatial] += displacement / 2
+            updated_pos = calculator.compute(system_pos)
+
+            system_neg = copy(system)
+            system_neg._positions[atom_i][spatial] -= displacement / 2
+            updated_neg = calculator.compute(system_neg)
```

Comment:
I am bit puzzled here. Updating the values does not work for whatever reasons. Also, the style of the code, especially `system_pos._positions` does only work with the SystemForTests object and not for `ase` or other objects.

Any ideas how to fix and improve without touching a lot of files?

Reply:
Have you tried a deepcopy? Also, you could make this function require an `ase.Atoms` object and rely on it to do modifications. Or even write a new small implementation of the system class just for these tests

Reply:
Deepcopy fails because parts of the c-code which stems from the baseclass can not be pickled.

We could do ase and only test these with all-deps test case. Should be fine.